src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java

```diff
@@ -813,10 +596,16 @@ private void finish(@Nonnull Result r, @CheckForNull Throwable t) {
                     Functions.printStackTrace(t, getListener().getLogger());
                 }
                 getListener().finished(getResult());
-                getListener().closeQuietly();
+                if (listener instanceof AutoCloseable) {
+                    try {
+                        ((AutoCloseable) listener).close();
+                    } catch (Exception x) {
+                        LOGGER.log(Level.WARNING, "could not close build log for " + this, x);
+                    }
+                }
+                listener = null;
             }
-            logsToCopy = null;
-            saveWithoutFailing();
+            saveWithoutFailing(); // TODO useless if we are inside a BulkChange
```

Comment:
Yeah I don't consider this a blocker, just something to keep in mind if we see issues with Pipelines not completing after this change. 

My initial thought for ensuring completion was to add a new `WorkflowRun#save` overload that takes a new `force` parameter, and if that is true then when `#save` returns early because of a bulk change, it sets a new transient `WorkflowRun.needsToSave` field to `true` (the field would be set to false at the end of `#save`) and schedules a background task to call `#save` again in a few seconds if that field is not `false`. Either way I think we can leave this for now and come back to it if we see any problems.

Reply:
> a new transient `WorkflowRun.needsToSave` field to `true` (the field would be set to false at the end of `#save`) and schedules a background task to call `#save` again in a few seconds if that field is not `false`

Should not be necessary if `BulkChange` is being used correctly with a `finally` block: sooner or later, in this thread, you are `save`ing. I have not followed the issues in https://github.com/jenkinsci/workflow-cps-plugin/pull/234 closely and perhaps there is an outstanding bug in `workflow-cps` in this regard.

At any rate, the behavior here is: whenever we do first successfully save `build.xml` after the end of the build (i.e., we are for the first time writing `<completed>true</completed>`), `AsynchronousExecution.completed` will be called. Previously this was done at a more random time (up to 3s later IIRC) in `copyLogsTask`.

Note that

> Pipelines not completing

is not likely what you would observe: the `AsynchronousExecution` corresponds only to the display of an unnumbered (flyweight) “executor slot” under **master** in the **Build Executor Status** widget, which is usually hidden (only rendered while the CPS VM thread is active). It has no effect on whether the build is considered completed for any other purpose, such as the **Build History** widget etc.

Reply:
Yes, I should clarify that I mean flyweight executors hanging around after the build completes rather than the build completion itself. The issue I am thinking of ([JENKINS-45571](https://issues.jenkins-ci.org/browse/JENKINS-45571)) was noticed by people looking at API output and seeing OneOffExecutors for long-completed jobs (which seems somewhat harmless though misleading), and by people using the Throttle Concurrent Builds Plugin where the flyweights count towards concurrent builds (not so harmless).

Reply:
> flyweights count towards concurrent builds

If true, that would be a bug in that plugin I think.

Reply:
> The issue I am thinking of ([JENKINS-45571](https://issues.jenkins-ci.org/browse/JENKINS-45571))

It does not sound like there is a definite way to reproduce that (?) but I see this JIRA comment:

> the build completes (`WorkflowRun#finish` is called), but the `copyLogsTask` never gets invoked or fails, since that is what actually removes the `FlyWeightTask` and kills the `OneOffExecutor`

This PR of course rewrites that logic, again with an explanation in 18d78f305a4526af9cdf3a7b68eb9caf97c7cfbc. We are no longer relying on the `copyLogs` system, which IIRC could sometimes get blocked on thread pool starvation, though perhaps this was bypassed by `GraphL` calling `copyLogs` directly. At any rate, now that call is guaranteed after `WorkflowRun.finish` sets `completed = true` so long as the `build.xml` gets saved _sometime_.

As to when that happens, since `finish` is not called from inside a `GraphListener.Synchronous`, the change in https://github.com/jenkinsci/workflow-cps-plugin/pull/234 should have no effect here AFAICT: with or without that, we are calling `BulkChange.commit` in a `finally`-block, which should result in `save` proceeding. So the only way you have a `FutureImpl` declining to consider the build complete (until the next restart) is if `build.xml` in fact still claims that the build is _not_ complete.

Reply:
Yes, there is no way to reproduce that issue that we know of, and that comment was our best hypothesis of the root cause. Like I said before, I am fine with the change, as it seems much easier to diagnose any issues with this implementation than it would be with `copyLogs`, and we can always revisit it if needed.

(Re: throttle-concurrent-builds: I think that behavior is intentional based on [a comment and the history](https://github.com/jenkinsci/throttle-concurrent-builds-plugin/blame/521b885be274b9a4fba13780dfaeb94891539051/src/main/java/hudson/plugins/throttleconcurrents/ThrottleQueueTaskDispatcher.java#L486), but I have not looked into it deeply.)

Reply:
The dangling OneOffExecutors have practical and irritating implications for tests, memory use, and garbage-collectibility of build records too.

The AsynchronousExecution should be terminated at the time we mark the completed flag. 

Integrating this aspect with the already complex (and unfortunately still somewhat failure-prone) persistence model is an exceedingly poor idea.

Reply:
> The AsynchronousExecution should be terminated at the time we mark the completed flag.

That was what I did originally. It did not work well for reasons explained in 18d78f305a4526af9cdf3a7b68eb9caf97c7cfbc. You need to wait until the state is actually written.

Reply:
Looking at that commit in depth, this choice makes *more* sense -- you basically wanted to guarantee it actually was saved before killing the AsynchronousExecution due to restarts.  But the catch there is we can't *guarantee* saves will work (disk failure, etc etc), and we do want to guarantee the AsynchronousExecution is completed when the build is. 

The on-restart logic is handled by #104 and that should do the bulk of needed cleanup for us if the persist-through-to-disk fails.

Reply:
We handle cases where the state is not written now though (#104) on restart. 

I'd say that if you really want to you could do something like this in `finish`:

```
try{
   if (BulkChange.contains(this)) {
         BulkChange.commit(this);
   }
   saveWithoutFailing()
} finally {
    `AsynchronousExecution.complete()`
}
```

(Note that I am writing from rough memory here, so I'm sure there are subtle typos in method signatures above, but you all get the gist).

Reply:
To give the perspective of a Jenkins user, I was affected by the abovementioned issue of flyweight executors hanging around after completed builds. I use the Throttle Concurrent Builds Plugin, and the flyweight executors counted towards the concurrent builds throttle (preventing new builds from running). This took some time to debug and delayed the release of one of my company’s products.

If upgrading to a build with JEP-210 means there is a risk of this type of problem reoccurring, I would be very hesitant to upgrade. As a Jenkins administrator, I am primarily focused on delivering a stable CI solution to the other engineers at my company (which is why I always use LTS releases of Jenkins core). If JEP-210 carries some degree of risk, perhaps this could be explicitly called out in the plugin release notes or update sites (with a stable version provided as an alternative).

Reply:
@basil Definitely appreciate the comment here to give some context on why this is important -- I swear we worked together to pin down some of the prior issues, and that's part of why this particular issue was in mind.  In this case, I'm working on a patch to amend this branch which provides stronger protection against the issue you mention.

That said, this release will carry extra risks simply because it's such a major change, despite our best efforts to harden it against bugs.  I'm not sure if we've finalized the release plan yet, since it might go out as a beta first (I hope this will be the case).

@jglick I'd encourage these log-handling-rewrite releases carry an extra warning in the changelog -- bolded and underlined preferably -- because we all know that despite our best efforts there is a higher than normal risk of regressions since this change touches a lot of sensitive pieces.